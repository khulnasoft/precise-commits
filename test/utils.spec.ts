import { TestBed, readFixtures } from './test-utils';
import {
  getDiffForFile,
  resolveNearestGitDirectoryParent,
} from '../src/git-utils';
import {
  extractLineChangeData,
  calculateCharacterRangesFromLineChanges,
} from '../src/utils';

const fixtures = readFixtures();
let testBed: TestBed;

describe('utils', () => {
  describe('extractLineChangeData()', () => {
    beforeAll(() => {
      testBed = new TestBed();
    });

    fixtures.forEach(fixture => {
      it(fixture.fixtureName, () => {
        testBed.prepareFixtureInTmpDirectory(fixture);
        const tmpFile = testBed.getTmpFileForFixture(fixture);
        const diff = getDiffForFile(
          resolveNearestGitDirectoryParent(tmpFile.directoryPath),
          tmpFile.path,
          tmpFile.initialCommitSHA,
          tmpFile.updatedCommitSHA,
        );
        const lineChangeData = extractLineChangeData(diff);
        expect(lineChangeData).toMatchSnapshot();
      });
    });
  });

  describe('calculateCharacterRangesFromLineChanges()', () => {
    beforeAll(() => {
      testBed = new TestBed();
    });

    fixtures.forEach(fixture => {
      it(fixture.fixtureName, () => {
        testBed.prepareFixtureInTmpDirectory(fixture);
        const tmpFile = testBed.getTmpFileForFixture(fixture);
        const diff = getDiffForFile(
          resolveNearestGitDirectoryParent(tmpFile.directoryPath),
          tmpFile.path,
          tmpFile.initialCommitSHA,
          tmpFile.updatedCommitSHA,
        );
        const lineChangeData = extractLineChangeData(diff);
        const characterRanges = calculateCharacterRangesFromLineChanges(
          lineChangeData,
          fixture.stagedContents,
        );
        expect(characterRanges).toMatchSnapshot();
      });
    });
  });
});
