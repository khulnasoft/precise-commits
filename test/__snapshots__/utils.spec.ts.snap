// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils calculateCharacterRangesFromLineChanges() brand-new-json-file 1`] = `
Array [
  Object {
    "rangeEnd": 19,
    "rangeStart": 0,
  },
]
`;

exports[`utils calculateCharacterRangesFromLineChanges() brand-new-jsx-file 1`] = `
Array [
  Object {
    "rangeEnd": 661,
    "rangeStart": 0,
  },
]
`;

exports[`utils calculateCharacterRangesFromLineChanges() empty-javascript-file 1`] = `
Array [
  Object {
    "rangeEnd": 53,
    "rangeStart": 0,
  },
]
`;

exports[`utils calculateCharacterRangesFromLineChanges() formatting-would-affect-total-line-count 1`] = `
Array [
  Object {
    "rangeEnd": 75,
    "rangeStart": 62,
  },
  Object {
    "rangeEnd": 130,
    "rangeStart": 105,
  },
]
`;

exports[`utils calculateCharacterRangesFromLineChanges() multiple-updated-lines-javascript-file 1`] = `
Array [
  Object {
    "rangeEnd": 127,
    "rangeStart": 32,
  },
  Object {
    "rangeEnd": 171,
    "rangeStart": 133,
  },
]
`;

exports[`utils calculateCharacterRangesFromLineChanges() single-updated-line-typescript 1`] = `
Array [
  Object {
    "rangeEnd": 29,
    "rangeStart": 14,
  },
]
`;

exports[`utils extractLineChangeData() brand-new-json-file 1`] = `
Object {
  "additions": Array [
    Object {
      "noOfLines": 3,
      "start": 1,
    },
  ],
  "removals": Array [],
}
`;

exports[`utils extractLineChangeData() brand-new-jsx-file 1`] = `
Object {
  "additions": Array [
    Object {
      "noOfLines": 21,
      "start": 1,
    },
  ],
  "removals": Array [],
}
`;

exports[`utils extractLineChangeData() empty-javascript-file 1`] = `
Object {
  "additions": Array [
    Object {
      "noOfLines": 2,
      "start": 1,
    },
  ],
  "removals": Array [],
}
`;

exports[`utils extractLineChangeData() formatting-would-affect-total-line-count 1`] = `
Object {
  "additions": Array [
    Object {
      "noOfLines": 2,
      "start": 5,
    },
    Object {
      "noOfLines": 2,
      "start": 10,
    },
  ],
  "removals": Array [
    Object {
      "noOfLines": 1,
      "start": 5,
    },
  ],
}
`;

exports[`utils extractLineChangeData() multiple-updated-lines-javascript-file 1`] = `
Object {
  "additions": Array [
    Object {
      "noOfLines": 1,
      "start": 2,
    },
    Object {
      "noOfLines": 1,
      "start": 6,
    },
  ],
  "removals": Array [
    Object {
      "noOfLines": 1,
      "start": 2,
    },
  ],
}
`;

exports[`utils extractLineChangeData() single-updated-line-typescript 1`] = `
Object {
  "additions": Array [
    Object {
      "noOfLines": 1,
      "start": 2,
    },
  ],
  "removals": Array [
    Object {
      "noOfLines": 1,
      "start": 2,
    },
  ],
}
`;
